primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Trait definitions for a Merkle tree gadget.
+
+use crate::merkle_tree::MerkleTreeScheme;
+use ark_ff::PrimeField;
+use jf_relation::{errors::CircuitError, BoolVar, Variable};
+
+mod rescue_merkle_tree;
+
+/// Circuit variable for a leaf element.
+#[derive(Debug, Clone)]
+pub struct LeafVar {
+    /// Position of the leaf element in the MT. Serves as UID.
+    pub uid: Variable,
+    /// The value of the leaf element.
+    pub elem: Variable,
+}
+
+/// Gadget for a Merkle tree
+///
+/// # Examples
+///
+/// ```
+/// use ark_bls12_377::Fq;
+/// use jf_primitives::circuit::merkle_tree::MerkleTreeGadget;
+/// use jf_relation::{Circuit, PlonkCircuit};
+/// use jf_primitives::merkle_tree::{prelude::RescueMerkleTree, MerkleTreeScheme, MerkleCommitment};
+///
+/// let mut circuit = PlonkCircuit::<Fq>::new_turbo_plonk();
+/// // Create a 3-ary MT, instantiated with a Rescue-based hash, of height 1.
+/// let elements = vec![Fq::from(1_u64), Fq::from(2_u64), Fq::from(100_u64)];
+/// let mt = RescueMerkleTree::<Fq>::from_elems(1, elements).unwrap();
+/// let expected_root = mt.commitment().digest();
+/// // Get a proof for the element in position 2
+/// let (_, proof) = mt.lookup(2).expect_ok().unwrap();
+///
+/// // Circuit computation with a MT
+/// let leaf_var = circuit.create_leaf_variable(Fq::from(2_u64), Fq::from(100_u64)).unwrap();
+/// let path_vars = circuit.create_membership_proof_variable(&proof).unwrap();
+/// let root_var = circuit.create_root_variable(expected_root).unwrap();
+/// circuit.enforce_merkle_proof(leaf_var, path_vars, root_var).unwrap();
+/// assert!(circuit.check_circuit_satisfiability(&[]).is_ok());
+/// ```
+pub trait MerkleTreeGadget<F, M>
+where
+    F: PrimeField,
+    M: MerkleTreeScheme,
```

Comment:
why does this trait have to be generic over both `F` and `M`?

Reply:
Your suggestion below is exactly the reason why it needs F right now. `RescueMerkleTree` defines the Index as `u64`, and that prevents an implementation `MerkleTreeGadget` for this specific MT.

I originally didn't want to touch the definition of `RescueMerkleTree`, but indeed it will make the interface even cleaner. We anyway cast the u64 index to `F` when using it (e.g. for hashing a leaf).